github_graphql.go

```diff
@@ -0,0 +1,70 @@
+package resource
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"time"
+
+	"github.com/google/go-github/v32/github"
+	"github.com/shurcooL/githubv4"
+)
+
+func (g *GitHubClient) listReleasesV4() ([]*github.RepositoryRelease, error) {
+	if g.clientV4 == nil {
+		return nil, errors.New("github graphql is not been initialised")
+	}
+	var listReleases struct {
+		Repository struct {
+			Releases struct {
+				Edges []struct {
+					Node struct {
+						ReleaseObject
+					}
+				} `graphql:"edges"`
+				PageInfo struct {
+					EndCursor   githubv4.String
+					HasNextPage bool
+				} `graphql:"pageInfo"`
+			} `graphql:"releases(first:$releasesCount, after: $releaseCursor, orderBy: {field: CREATED_AT, direction: DESC})"`
+		} `graphql:"repository(owner:$repositoryOwner,name:$repositoryName)"`
+	}
+
+	vars := map[string]interface{}{
+		"repositoryOwner": githubv4.String(g.owner),
+		"repositoryName":  githubv4.String(g.repository),
+		"releaseCursor":   (*githubv4.String)(nil),
+		"releasesCount":   githubv4.Int(100),
+	}
+
+	var allReleases []*github.RepositoryRelease
+	for {
+		if err := g.clientV4.Query(context.TODO(), &listReleases, vars); err != nil {
+			return nil, err
+		}
+
+		for _, r := range listReleases.Repository.Releases.Edges {
+			r := r
+			publishedAt, _ := time.ParseInLocation(time.RFC3339, r.Node.PublishedAt.Time.Format(time.RFC3339), time.UTC)
+			createdAt, _ := time.ParseInLocation(time.RFC3339, r.Node.CreatedAt.Time.Format(time.RFC3339), time.UTC)
+			releaseID, _ := strconv.ParseInt(r.Node.ID, 10, 64)
```

Comment:
This always seems to give `0`. I noticed that the ID given by the graphql API is actually a base64 encoded string that ends up looking like `07:Release33356981` when decoded.

I think the ID is intended to be an opaque string, but I'm hesitant to break compatibility with previous versions of the github-release resource by getting rid of the ID field.

What do you think about trying to parse it out? e.g.

```go
var releaseID int64
decodedID, err := base64.StdEncoding.DecodeString(r.Node.ID)
if err == nil {
	releaseID, _ = strconv.ParseInt(strings.TrimPrefix(decodedID, "07:Release"), 10, 64)
}
```

It feels brittle, and probably is, but at least it'll maintain backward compatibility (until GitHub decides to change the ID format or something, but we can figure that out if the time comes).

WDYT?

Reply:
Strange when I did the graphql explorer it was returning the correct id, just wondering if this is something connected to the library used for graphql, let me check deeper

Reply:
@aoldershaw not sure about this approach, are we sure that's id will be compatible with v3? I mean that's id from graphql will match id for api v3 in this case. I found that' we have method get a release, which accepts id as input GetRelease(id).

Reply:
> are we sure that's id will be compatible with v3?

It'd be nice if it were documented/explicitly supported, but the IDs seem to line up. For the latest concourse release (6.7.1), I get a GraphQL ID of `MDc6UmVsZWFzZTMzMzU2OTgx`, which decodes to `07:Release33356981`. The v3 API gives `33356981`.

> I found that' we have method get a release, which accepts id as input GetRelease(id)

Ah, good point. If we do go with the approach of trying to parse out the ID, I would error out in that case - e.g.

```go
decodedID, err := base64.StdEncoding.DecodeString(r.Node.ID)
if err != nil {
	return nil, err
}
releaseID, err := strconv.ParseInt(strings.TrimPrefix(decodedID, "07:Release"), 10, 64)
if err != nil {
	return nil, err
}
```

so at least we know that our parsing is causing issues (e.g. in case GitHub changes their IDs or something).

But I feel there has to be a better way...

Reply:
I suppose we could `GetReleaseByTag` instead of by ID - not sure if the tag can change for a given ID, or vice versa

Reply:
I was thinking to use regexp to parse id, if it will fail we can try to use a tag as a backup solution?

Reply:
Yeah, regexp makes sense, as long as it's anchored to the end of the string. If it's not too much work, it might be a good idea to fallback to tag as you suggest